### PR TITLE
Refactoring.

### PR DESCRIPTION
**Pull request by Htocki**

### CODE REVIEW

File: lib/rational/CMakeLists.txt [hunk 1]
```diff
@@ -30,4 +30,5 @@ target_include_directories(${PROJECT_NAME}
         ${CMAKE_CURRENT_SOURCE_DIR}/src
 	)
 
+add_subdirectory ("bin")
```
Comment (sialiuk) [inline lib/rational/CMakeLists.txt]: Why do we need to create bin folder?

File: lib/rational/bin/Main.cpp [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#include "Rational.h"
+
+int main() {
+	Rational default;
```
Comment (sialiuk) [inline lib/rational/bin/Main.cpp]: Don't use c++ keywords as names in your code

File: lib/rational/include/Rational.h [hunk 3]
```diff
@@ -1,33 +1,36 @@
 #pragma once
 
-
+#include <iostream>
```
Comment (sialiuk) [inline lib/rational/include/Rational.h]: Replace with #include <iosfwd>
Please move #include <iostream> to cpp file

File: lib/rational/include/Rational.h [hunk 4]
```diff
@@ -1,33 +1,36 @@
...
 	Rational(int numerator, int denominator);
 
-	Rational operator+(const Rational& other);
-	Rational operator-(const Rational& other);
-	Rational operator*(const Rational& other);
-	Rational operator/(const Rational& other);
+	Rational operator+=(const Rational other);
+	Rational operator-=(const Rational other);
+	Rational operator*=(const Rational other);
+	Rational operator/=(const Rational other);
```
Comment (sialiuk) [inline lib/rational/include/Rational.h]: Should return Rational&

File: lib/rational/include/Rational.h [hunk 5]
```diff
@@ -1,33 +1,36 @@
...
+
+	double ToDouble();
 
-	double toDouble();
+	friend std::ostream& operator<<(
+		std::ostream& os,
+		const Rational
```
Comment (sialiuk) [inline lib/rational/include/Rational.h]: Don't avoid writing input parameters names

File: lib/rational/src/Rational.cpp [hunk 6]
```diff
@@ -1,61 +1,84 @@
 #include "Rational.h"
 
-
-
-Rational::Rational() {}
+namespace {
+	int Gcd(int a, int b);
+	int Lcm(int a, int b);
+}
```
Comment (sialiuk) [inline lib/rational/src/Rational.cpp]: You can implement functions right here

File: lib/rational/include/Rational.h [hunk 7]
```diff
@@ -1,33 +1,36 @@
...
 
 private:
-	Rational reduce() const;
+	auto Print(std::ostream& os) const ->std::ostream&;
+	Rational Reduce() const;
 
-	int numerator = 1;
-	int denominator = 1;
+	int numerator = { 1 };
+	int denominator = { 1 };
```
Comment (sialiuk) [inline lib/rational/include/Rational.h]: Use prefix or postfix for members names

File: lib/rational/src/Rational.cpp [hunk 8]
```diff
@@ -1,61 +1,84 @@
...
+	);
+	temp.Reduce();
+	this->numerator = temp.numerator;
+	this->denominator = temp.denominator;
+	return *this;
+}
+
+double Rational::ToDouble() {
+	return static_cast<double>(numerator) /
+		static_cast<double>(denominator);
```
Comment (sialiuk) [inline lib/rational/src/Rational.cpp]: Don't use static_cast

File: lib/rational/src/Rational.cpp [hunk 9]
```diff
@@ -1,61 +1,84 @@
...
+std::ostream& operator<<(std::ostream& os, const Rational other) {
+	return other.Print(os);
 }
 
-bool operator==(const Rational& left, const Rational& right) {
-	Rational a = left.reduce();
-	Rational b = right.reduce();
+bool operator==(const Rational lhs, const Rational rhs) {
+	Rational a = lhs.Reduce();
+	Rational b = rhs.Reduce();
```
Comment (sialiuk) [inline lib/rational/src/Rational.cpp]: Return a.numerator == b.numerator && a.denominator == b.denominator;

File: lib/rational/src/Rational.cpp [hunk 10]
```diff
@@ -64,29 +87,61 @@ bool operator==(const Rational& left, const Rational& right) {
...
 
-Rational Rational::reduce() const {
-	int divisor = gcd(this->numerator, this->denominator);
+auto Rational::Print(std::ostream& os) const->std::ostream& {
+	os << this->numerator << "/" << this->denominator;
+	return os;
+}
+
+Rational Rational::Reduce() const {
+	int divisor = Gcd(this->numerator, this->denominator);
```
Comment (sialiuk) [inline lib/rational/src/Rational.cpp]: const auto divisor

**Review by sialiuk — CHANGES_REQUESTED** [review]

File: lib/rational/include/Rational.h (hunk 3 above)
Comment (Htocki) [reply lib/rational/include/Rational.h]: What for?

File: lib/rational/include/Rational.h (hunk 7 above)
Comment (Htocki) [reply lib/rational/include/Rational.h]: What do you have in mind?

File: lib/rational/include/Rational.h (hunk 3 above)
Comment (sialiuk) [reply lib/rational/include/Rational.h]: It's better to use forward declarations, isn't it?

File: lib/rational/include/Rational.h (hunk 7 above)
Comment (sialiuk) [reply lib/rational/include/Rational.h]: Members should be named like this
m_numerator or numerator_;

**Review by sialiuk — CHANGES_REQUESTED** [review]

File: lib/rational/include/Rational.h (hunk 3 above)
Comment (Htocki) [reply lib/rational/include/Rational.h]: Why is it better?

File: lib/rational/src/Rational.cpp [hunk 11]
```diff
@@ -60,83 +67,78 @@ Rational& Rational::operator-=(const Rational other) {
...
-	return Rational{
-		this->numerator / divisor,
-		this->denominator / divisor
-	};
+void Rational::Reduce() {
+	const auto divisor = Gcd(numerator_, denominator_);
+	Rational temp = *this;
+	temp.numerator_ /= divisor;
+	temp.denominator_ /= divisor;
+	*this = temp;
```
Comment (sialiuk) [inline lib/rational/src/Rational.cpp]: It looks like we don't need to create a temp object.
You only need to change members
numerator_ /= divisor;
denominator_ /= divisor;

**Review by sialiuk — CHANGES_REQUESTED** [review]

**Review by sialiuk — APPROVED** [review]

**Pull request merged**